cmd/tree-walk.go

```diff
@@ -123,7 +114,9 @@ func filterListEntries(bucket, prefixDir string, entries []string, prefixEntry s
 
 	// Sort again after removing trailing "/" for objects as the previous sort
 	// does not hold good anymore.
-	sort.Strings(entries)
+	if !sort.StringsAreSorted(entries) {
```

Comment:
It looks like this check is useless, `entries` slice is always modified at this point

Reply:
Not if the comment is correct. @vadmeste

Reply:
```
// Sort again after removing trailing "/" for objects as the previous sort
// does not hold good anymore.
```

This means that the sorting is not kept

Reply:
@vadmeste Ah, that is what you meant.

Checking if a slice is sorted is much faster than actually sorting it. And no, the sort order isn't that likely to have changed, only if you have objects with names that overlaps with a prefix AFAICT.

Reply:
```
func BenchmarkSort(b *testing.B) {
	var x []string
	rng := rand.New(rand.NewSource(0))
	var tmp [32]byte
	for i := 0; i < 10000; i++ {
		rng.Read(tmp[:])
		x = append(x, hex.EncodeToString(tmp[:]))
	}
	sort.Strings(x)
	b.Run("sort", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sort.Strings(x)
		}
	})
	b.Run("aresorted", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			sort.StringsAreSorted(x)
		}
	})
}
```
Output:

```
BenchmarkSort
BenchmarkSort/sort
BenchmarkSort/sort-32         	    1052	   1081749 ns/op
BenchmarkSort/aresorted
BenchmarkSort/aresorted-32    	   13028	     91035 ns/op
```

Reply:
2 orders of magitude faster if slice is already sorted.

Reply:
Now that I think about it @klauspost I think @vadmeste is saying that sort.Strings() is needed in almost all cases > 50% so it would mean that the happy case is additionally taxed to check if it's already sorted, which is an added overhead.

You would have to take this into account.

Reply:
If we look at the flow here I think sort.Strings() is only needed on the slice if the delayIsLeaf check is false, that is rare and only happens with conflicting prefixes @klauspost 

So it may be okay to have the check but then if we did hit the applications which have such common prefixes - check why delayIsLeaf was implemented. 

So in essence we would pay an additional CPU cycle for checking if the slice is sorted which may not needed @klauspost